File: student-geolocation-python/student_geolocation/geolocation_simulation.py
from models.classroom import ClassRoom
from models.student import Student


class GeolocationSimulation:
    """Earth class to find students in course"""

    def students_in_classes(self, students, class_rooms):
        """
        Function to verify if a bunch of students is inside of a list of class
        room.

        @param students : is a students list -> [Student]
        @param class_rooms : is a class rooms list -> [ClassRoom]

        @return: list with all students found inside a course
        @rtype : [Student]
        """
        students_found = []
        for class_room in class_rooms:
            for student in students:
                if class_room.is_student_in_class_room(student.position):
                    students_found.append(student)

        return students_found

    def student_clusters_in_classes(self, students, class_rooms):
        """
        Function to find and return class rooms that has at least two students
        inside.

        @param students : is a students list -> [Student]
        @param class_rooms : is a class rooms list -> [ClassRoom]

        @return: a class room list whit has at least two students
        @rtype : [ClassRoom]
        """
        # Closure function to verify if the student is inside of classRoom
        def verify(student, clas):
            if not hasattr(clas, 'students'):
                clas.students = []
            if clas.is_student_in_class_room(student.position):
                clas.students.append(student)
            return clas

        for student in students:
            class_rooms = map((lambda c: verify(student, c)), class_rooms)

        # filter classRoom that has at least two students
        class_rooms = filter((lambda c: len(c.students) >= 2), class_rooms)
        return class_rooms

    def print_clouster_students(self, class_rooms):
        for class_room in class_rooms:
            print class_room.name + " has the following students:"
            for student in class_room.students:
                print student.name

    def example_simulation_one(self):
        class_rooms = [
            ClassRoom("Principles", 34.069140, -118.442689),
            ClassRoom("Sedimentary Petrology", 34.069585, -118.441878),
            ClassRoom("Introductory Psychobiology", 34.069742, 118.441312),
            ClassRoom("Art of Listening", 34.070223, -118.440193),
            ClassRoom("Art Hitory", 34.071528, -118.441211)
        ]

        students = [
            Student("John Wilson", 34.069149, -118.442639),
            Student("Jane Graham", 34.069601, -118.441862),
            Student("Pame Bam", 34.071513, -118.441181),
            Student("Bam Bam", 34.071501, -118.441170)
        ]

        self.print_clouster_students(
            self.student_clusters_in_classes(students, class_rooms))

File: student-geolocation-java/src/studentgeolocation/GeolocationSimulation.java
package studentgeolocation;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

public class GeolocationSimulation {

	public GeolocationSimulation() {

	}

	/**
	 * Function to verify if a bunch of students are inside of a list of courses
	 * 
	 * @param students   <code>List<Student></code> students list to be searching inside 
	 *   				 of list a classrooms. 
	 * @param classRooms <code>List<ClassRoom></code> classrooms list to verify if a students
	 *                   are inside.   
	 * @return           <code>List<Student></code> a list with all students found 
	 * 					 inside a course.
	 */
	public List<Student> studentsInClasses(List<Student> students, List<ClassRoom> classRooms) {
		List<Student> studentsFound = new ArrayList<>();
		for (ClassRoom classRoom : classRooms) {
			for (Student student : students) {
				if (classRoom.isStudentInClassRoom(student.getPosition())) {
					studentsFound.add(student);
				}
			}
		}
		return studentsFound;
	}

	public void printStudentsList(List<Student> students) {
		for (Student s : students) {
			System.out.println("name:" + s.getName() + " , lat:" + s.getPosition().getLatitude() + " ,long:"
					+ s.getPosition().getLongitude());
		}
	}

	public void exampleSimulationOne() {
		List<Student> students = new ArrayList<>(Arrays.asList(
				new Student("John Wilson", 34.069149, -118.442639), 
				new Student("Jane Graham", 34.069601, -118.441862),
				new Student("Pam Bam", 34.071513, -118.441181)));

		List<ClassRoom> classRooms = new ArrayList<>(Arrays.asList(
				new ClassRoom("Principles of computational", 34.069140, -118.442689),
				new ClassRoom("Sedimentary Petrology", 34.069585, -118.441878),
				new ClassRoom("Introductory Psychobiology", 34.069742, 118.441312),
				new ClassRoom("Art of Listening", 34.070223, -118.440193),
				new ClassRoom("Art Hitory", 34.071528, -118.441211)));

		printStudentsList(studentsInClasses(students, classRooms));
	}

	public void exampleSimulationTwo() {
		List<Student> students = new ArrayList<>(Arrays.asList(
				new Student("John Wilson", 34.069849, -118.443539),
				new Student("Jane Graham", 34.069901, -118.441562),
				new Student("Pam Bam", 34.071523, -118.441171)));
		
		List<ClassRoom> classRooms = new ArrayList<>(Arrays.asList(
				new ClassRoom("Principles of computational", 34.069140, -118.442689),
				new ClassRoom("Sedimentary Petrology", 34.069585, -118.441878),
				new ClassRoom("Introductory Psychobiology", 34.069742, -118.441312),
				new ClassRoom("Art of Listening", 34.070223, -118.440193),
				new ClassRoom("Art Hitory", 34.071528, -118.441211)));

		printStudentsList(studentsInClasses(students, classRooms));
	}

}
